File: practices/chapter5-stack/ch_5_5_factorial.js
const readline = require('readline')

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
})

class Factorial {
  constructor(n=1) {
    this.fact = 0
    this.getFactorial(n)
  }

  printFactBeforeCalled(n) {
    console.log(`factorial(${n})called before ${n}! = ${this.fact}`)
  }

  printFactBeforeReturned(n) {
    console.log(`factorial(${n})called returned ${n}! = ${this.fact}`)
  }

  getFactorial(n) {
    if(n === 1) {
      this.printFactBeforeCalled(n)
      console.log('Arrive at recursion endding n = 1')
      this.fact = 1
      this.printFactBeforeReturned(n)
      return this.fact
    } else {
      this.printFactBeforeCalled(n)
      this.fact = n * this.getFactorial(n - 1)
      this.printFactBeforeReturned(n)
      return this.fact
    }
  }
}

function main() {
  rl.question('Input N of factorial: ', ans => {
    const fact = new Factorial(Number(ans))
    console.log(fact.fact)
    rl.close()
  })
}

main()

File: practices/chapter5-stack/ch_5_5_factorial.py
class Factorial():
  def __init__(self):
    self.fact = 0
  
  def print_fact_called_before(self, n):
    print('factorial(%d)called before %d! = %d' % (n, n, self.fact))

  def print_fact_called_returned(self, n):
    print('factorial(%d)called returned: %d! = %d' % (n, n, self.fact))

  def factorial(self, n):
    if n == 1:
      self.print_fact_called_before(n)
      print('Arrive at recursion endding n = 1.')
      self.fact = 1
      self.print_fact_called_returned(n)
      return self.fact
    else:
      self.print_fact_called_before(n)
      self.fact = n * self.factorial(n - 1)
      self.print_fact_called_returned(n)
      return self.fact

def main():
  n = eval(input('Input factor n: '))
  # n = 5
  print(n, 'of factorial is = ', Factorial().factorial(n))

if __name__ == "__main__":
    main()

